External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h

```diff
@@ -0,0 +1,507 @@
+  class ThreadPoolAllocatorMMap final : public IntrusiveThreadPoolAllocator {
+    public:
+      ThreadPoolAllocatorMMap() = default;
+
+      virtual ~ThreadPoolAllocatorMMap() {
+        // Cleanup any buffers that are remaining
+        for (auto it : OwnedBuffers) {
+          Free(it->Ptr, it->Size);
+          delete it;
+        }
+
+        for (auto it : NonOwnedBuffers) {
+          Free(it->Ptr, it->Size);
+          delete it;
+        }
+        OwnedBuffers.clear();
+        NonOwnedBuffers.clear();
+      }
+
+    private:
+      void *Alloc(size_t Size) override {
+        return FEXCore::Allocator::mmap(0, Size,
+          PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+      }
+
+      void Free(void* Ptr, size_t Size) override {
+        FEXCore::Allocator::munmap(Ptr, Size);
+      }
+  };
+
+  /**
+   * @brief Wrapper around the pool allocator for delayed pool reclaiming
+   *
+   * This is expected to be used in high frequency buffer temporary usage.
+   * Instead of quickly unclaiming and reclaiming the buffer while the the code is hot,
+   * This instead will do the cheap operation of disowning the buffer until the code path cools down enough.
+   * Once the code path stops disowning the codepath more times than `PeriodFrequency` during `PeriodMS` then
+   * it will immediately unclaim.
+   *
+   * Implications:
+   *   - The object will always be claimed for at *least* `PeriodFrequency`
+   *   - The object will still *always* be disowned after each temporary use
+   *     - This allows the pool allocator to reclaim a buffer from a sleeping thread
+   *
+   * Performance characteristics:
+   *  - Disowning is cheap.
+   *    - Last-used timestamp update
+   *    - atomic_bool clear to signify it is disowned
+   *
+   *  - Reowning is relatively cheap (When buffer is still owned).
+   *    - atomic_bool load to check if the object is still owned
+   *      - atomic<uint32_t> CAS to change the object to `OWNED` state
+   *        - Resolves a race condition where the `Allocator` can be in the process of reclaiming the buffer from the client
+   *      - Last-used timestamp update
+   *      - atomic_bool<relaxed> set to signify owned
+   *      - atomic<uint32_t> set to change object to `OWNED` state
+   *    - When object isn't owned, then allocate a new buffer from the pool
+   *
+   *  - Unclaiming is fairly costly
+   *    - Requires owning a mutex, shared between all threads using the `Allocator`
+   *    - Updating two std::list containers to give the ownership back to the `Allocator`
+   *
+   *  - Claiming is very costly
+   *    - Requires owning a mutex, shared between all threads using the `Allocator`
+   *    - Scans two std::list containers to find the best fit buffer
+   *    - Or allocates another buffer when that fails
+   *    - Frees stale buffers opportunistically
+   */
+  template<typename Type, size_t PeriodMS, size_t PeriodFrequency>
+  class FixedSizeDelayedObjectReclaimer final {
```

Comment:
How about we name this from `FixedSizeDelayedObjectReclaimer` to `FixedSizePooledAllocation` for clarity?

Reply:
Resolved.